ouroboros-network/src/Ouroboros/Network/AnchoredFragment/Completeness.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveAnyClass #-}
+
+module Ouroboros.Network.AnchoredFragment.Completeness (
+    AnnotatedAnchoredFragment (..)
+  , FragmentCompleteness (..)
+  , emptyAnnotatedAnchoredFragment
+  , isFragmentComplete
+  ) where
+
+import           GHC.Generics (Generic)
+import           NoThunks.Class (NoThunks)
+
+import           Ouroboros.Network.AnchoredFragment
+import           Ouroboros.Network.Block
+import           Cardano.Slotting.Slot
+
+-- | BlockFetch and ChainSync operate with 'AnchoredFragment's when dealing with
+-- candidate chains. For Ouroboros Genesis, we need to know whether or not the
+-- fragment represents the end of the remote chain, to make judgments about
+-- whether or not our perceived density is indeed the real density.
+--
+-- Every fragment returned by ChainSync will be coupled with one of these tags,
+-- specifying if the ChainSync client thinks that the fragment is indeed
+-- complete. This flag will be used by the prefix selection algorithm to filter
+-- appropriate candidates to be provided to BlockFetch.
+--
+-- Note that completeness of a fragment is a local property, dependent on the
+-- current state of the ChainSync protocol, and based on the information
+-- provided by the ChainSync server.
+data FragmentCompleteness = FragmentComplete | FragmentIncomplete
+  deriving (Show, Generic, Eq, NoThunks)
+
+-- | Return a 'FragmentCompleteness' for the given fragment assuming the
+-- provided tip is the one announced by the remote peer.
+--
+-- This function will return 'Nothing' when the slot of the announced tip is
+-- smaller than the slot of the most recent block on the fragment. This should
+-- be interpreted in the ChainSync protocol as a misbehave of the server.
```

Comment:
I wonder if we should be encoding "the server is misbehaving" as `Nothing`. Doesn't this couple this part of the logic with the callers? So for instance if this function changes and returns `Nothing` in other scenarios we would be inadvertently breaking the clients of this function.

Reply:
Oh, and why should this be interpreted as the server misbehaving? :grin: